src/tqec/computation/open_graph.py

```diff
@@ -0,0 +1,220 @@
+"""Utilities for working with graphs with open ports."""
+
+from dataclasses import dataclass
+from functools import reduce
+from itertools import combinations
+from typing import Iterator
+
+import stim
+import networkx as nx
+
+from tqec.computation.block_graph import BlockGraph
+from tqec.computation.correlation import CorrelationSurface
+from tqec.computation.cube import YHalfCube, ZXCube
+from tqec.utils.enums import Basis
+from tqec.utils.exceptions import TQECException
+from tqec.utils.position import Direction3D
+
+
+@dataclass(frozen=True)
+class FilledGraph:
+    """A block graph without open ports that can be used for simulation.
+
+    Attributes:
+        graph: The block graph with all ports filled.
+        stabilizers: The external stabilizers simulatable on the filled graph.
+        observables: The correlation surfaces that can be used as logical observables
+            for the simulation. A list of independent correlation surfaces with the
+            area as small as possible is chosen.
+    """
+
+    graph: BlockGraph
+    stabilizers: list[str]
+    observables: list[CorrelationSurface]
+
+    def get_external_stabilizers(self) -> list[str]:
+        """Return the external stabilizers of the correlation surfaces."""
+        return [s.external_stabilizer_on_graph(self.graph) for s in self.observables]
+
+
+def fill_ports_for_minimal_simulation(graph: BlockGraph) -> list[FilledGraph]:
+    """Given a block graph with open ports, fill in the ports with the appropriate
+    cubes that will minimize the number of simulation runs needed for the complete
+    logical observable set.
+
+    Args:
+        graph: The block graph with open ports.
+
+    Returns:
+        A list of :class:`~tqec.computation.open_graph.FilledGraph` instances, each
+        containing a block graph with all ports filled and a set of correlation
+        surfaces that can be used as logical observables for the simulation on that
+        block graph.
+    """
+    num_ports = graph.num_ports
+    if num_ports == 0:
+        raise TQECException("The provided graph has no open ports.")
```

Comment:
Should this raise? At first glance it seems like
```suggestion
    if num_ports == 0:
        return FilledGraph(graph, [], [])
```
would be semantically valid.

Reply:
The `FilledGraph` includes the external stabilizer attribute that can only be computed before filling the ports. To be consistent, we should specify that even for zero-port input, but it requires some unnecessary complexity.